File: scripts/parity/parity.R
# D10 parity: run a downloaded bundle's query/*.sql in R and compare the lens table to the CSV the browser wrote.
# usage: Rscript scripts/parity/parity.R <unzipped bundle dir>
args <- commandArgs(TRUE); dir <- if (length(args)) args[1] else "."
setwd(dir); suppressPackageStartupMessages(library(DBI))
con <- dbConnect(duckdb::duckdb()); dbExecute(con, "INSTALL httpfs; LOAD httpfs")
run <- function(f) { sql <- paste(readLines(file.path("query", f)), collapse = "\n"); if (grepl("^\\s*(--.*\\n)*\\s*CREATE", sql)) dbExecute(con, sql) else dbGetQuery(con, sql) }
fs <- sort(list.files("query", pattern = "\\.sql$")); run(fs[1]); r <- run(fs[2])
app_csv <- list.files("data/summary", pattern = "\\.csv$", full.names = TRUE)[1]; app <- read.csv(app_csv)
key <- intersect(names(r), names(app))[1]
m <- merge(r, app, by = key, suffixes = c(".r", ".app"))
num <- intersect(names(r)[sapply(r, is.numeric)], names(app)[sapply(app, is.numeric)]); num <- setdiff(num, key)
maxd <- max(sapply(num, function(c) max(abs(m[[paste0(c, ".r")]] - m[[paste0(c, ".app")]]), na.rm = TRUE)))
ok <- nrow(r) == nrow(app) && nrow(m) == nrow(r) && maxd < 1e-9
cat(sprintf("%s: R rows %d · app rows %d · matched %d on %s · max |diff| over %s = %.3g → %s\n", fs[2], nrow(r), nrow(app), nrow(m), key, paste(num, collapse = ","), maxd, if (ok) "PARITY" else "MISMATCH"))
quit(status = if (ok) 0 else 1)

File: scripts/parity/parity.py
"""D10 parity: run a downloaded bundle's query/*.sql in Python and compare the lens table to the CSV the browser wrote.
usage: python scripts/parity/parity.py <unzipped bundle dir>"""
import glob, os, re, sys
import duckdb, pandas as pd
os.chdir(sys.argv[1] if len(sys.argv) > 1 else ".")
con = duckdb.connect(); con.execute("INSTALL httpfs; LOAD httpfs")
def run(f):
    sql = open(f"query/{f}").read()
    return con.execute(sql) if re.match(r"^\s*(--.*\n)*\s*CREATE", sql) else con.execute(sql).df()
fs = sorted(os.path.basename(f) for f in glob.glob("query/*.sql")); run(fs[0]); r = run(fs[1])
app = pd.read_csv(glob.glob("data/summary/*.csv")[0])
key = [c for c in r.columns if c in app.columns][0]
m = r.merge(app, on=key, suffixes=("_py", "_app"))
num = [c for c in r.columns if c != key and c in app.columns and pd.api.types.is_numeric_dtype(r[c]) and pd.api.types.is_numeric_dtype(app[c])]
maxd = max((m[f"{c}_py"] - m[f"{c}_app"]).abs().max() for c in num)
ok = len(r) == len(app) == len(m) and maxd < 1e-9
print(f"{fs[1]}: py rows {len(r)} · app rows {len(app)} · matched {len(m)} on {key} · max |diff| over {','.join(num)} = {maxd:.3g} → {'PARITY' if ok else 'MISMATCH'}")
sys.exit(0 if ok else 1)
